### PR TITLE
[wallet] Modify change identification to use hdsplit keypath

### PR DESCRIPTION
**Pull request by instagibbs**
Keypath data is a much more reliable way of detecting change outputs. 

There are a number of cases where the legacy detection fails, leading to odd results, such as a wallet rescan that results in missed results in `listtransactions` with self-sends.

### CODE REVIEW

**Comment by instagibbs** [comment]
I will add listtransaction test soon if people want.

**Comment by instagibbs** [comment]
pushed a test that covered a corner case this fixes

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1327,22 +1327,31 @@ CAmount CWallet::GetCredit(const CTxOut& txout, const isminefilter& filter) cons
...
-        if (!mapAddressBook.count(address))
-            return true;
+        // HD Split assumes keypath of m/.../1(')/k
+        if (CanSupportFeature(FEATURE_HD_SPLIT)) {
+            CKeyID key_id = GetKeyForDestination(*this, dest);
+            if (!key_id.IsNull()) {
+                auto it = mapKeyMetadata.find(key_id);
+                if (it != mapKeyMetadata.end()) {
+                    // Check for path
+                    if (it->second.hdKeypath.find("m/0'/1'/") != std::string::npos) {
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: Not for this PR, but we should probably factor out the static key path strings... maybe add a `IsChangeKeypath()`.

**Comment by jonasschnelli** [comment]
utACK 843e07cd5ffb4c9f8d73ef0e2aeed03303122a5b
Seems to not create any performance impacts.

**Review by jonasschnelli — APPROVED** [review]

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: Yes, I think we should be more systematic about our keypath string handling; would make it easier for people like me with various wallet forks and for Core as we move forward with additional derivation features.

**Comment by sipa** [comment]
I'm afraid this is not sufficient. It's possible to import change keys/addresses using `importmulti` (see the "internal" field); these aren't identified by having a change-like keypath. Furthermore, this approach will become even less reliable if it's possible to upgrade from pre-hd or pre-hdsplit wallets, as the old keys should remain being treated as change.

You could use being present in a change-like key path as a shortcut still and fall back to the other mechanism otherwise, but I doubt that's very useful.

**Comment by instagibbs** [comment]
@sipa actually that suggestion is just fine for my needs, even if you aren't too excited about it.

**Comment by instagibbs** [comment]
oh hmm, I see why now. I'll just carry it downstream for now.

**Pull request closed without merging**